### ISSUE TITLE
Deleted Logfiles still referenced

### ISSUE THREAD

**Issue by ulrichSchreiner** [timberio/vector#8101]
hi,

we are using vector as an agent in k8s. when the agent runs for a day or so, it leaks disk space because deleted logfiles are still referenced. whe doing `find /proc/*/fd -ls | grep  '(deleted)'` i get this (PID/3961658 == vector)
~~~
/proc/3961658/fd/40 -> /var/log/pods/elastic-...100b1-ff77-49f7-9794-6c1d066eceb3/elasticsearch/1.log.20210701-075620\ (deleted)
/proc/3961658/fd/42 -> /var/log/pods/elastic-...100b1-ff77-49f7-9794-6c1d066eceb3/elasticsearch/1.log.20210701-075943\ (deleted)
/proc/3961658/fd/45 -> /var/log/pods/elastic-...100b1-ff77-49f7-9794-6c1d066eceb3/elasticsearch/1.log.20210701-075338\ (deleted)
/proc/3961658/fd/46 -> /var/log/pods/elastic-...100b1-ff77-49f7-9794-6c1d066eceb3/elasticsearch/1.log.20210701-075036\ (deleted)
/proc/3961658/fd/50 -> /var/log/pods/elastic-...100b1-ff77-49f7-9794-6c1d066eceb3/elasticsearch/1.log.20210701-075015\ (deleted)
/proc/3961658/fd/51 -> /var/log/pods/elastic-...100b1-ff77-49f7-9794-6c1d066eceb3/elasticsearch/1.log.20210701-080326\ (deleted)
/proc/3961658/fd/52 -> /var/log/pods/elastic-...100b1-ff77-49f7-9794-6c1d066eceb3/elasticsearch/1.log.20210701-080347\ (deleted)
/proc/3961658/fd/54 -> /var/log/pods/elastic-...100b1-ff77-49f7-9794-6c1d066eceb3/elasticsearch/1.log.20210701-080448\ (deleted)
/proc/3961658/fd/55 -> /var/log/pods/elastic-...100b1-ff77-49f7-9794-6c1d066eceb3/elasticsearch/1.log.20210701-075257\ (deleted)
/proc/3961658/fd/56 -> /var/log/pods/elastic-...100b1-ff77-49f7-9794-6c1d066eceb3/elasticsearch/1.log.20210701-080528\ (deleted)
... (and much more)
~~~

the logfiles `1.log.YYYYMMDD-HHMMSS` are created by the docker daemon which has the opts:
~~~
"log-driver": "json-file",
"log-opts": {
  "max-size": "50m",
  "max-file": "3"
},
~~~
and the elastic process creates MANY logs. when i (re-)start the vector agent, everything is ok and i can look at it every hour. no problem. the rotated logfiles are created and vector releases them and the upper find command returns no problem. but after one day i do this `find` command and now the process with PID=3961658 (=vector) has these files open and so the disk space is not freed. 

i'm not sure why these files are opened by the agent, because i thought only files which ends in `*.log` are monitored? are this files open because they were named `1.log` and the rotation renamed them? in this case i expected the files to be closed.

EDIT: 
Our machines have some sort of buggy logging configuration. The docker daemon rotates the log files, but there is also a timerd-job which is run every 10 minutes
~~~
/var/log/pods/*/*/*.log {
  rotate 14
  copytruncate
  missingok
  notifempty
  compress
  maxsize 100M
  daily
  dateext
  dateformat -%Y%m%d-%s
  create 0644 root root
}
~~~
This is a known bug (https://github.com/gardener/gardener-extension-os-suse-chost/issues/6) because both rotation can (and will) touch the same file sometimes. It will be fixed in the future.
I don't know if this is problem leads to the error in vector so sometimes files descriptors are not closed. If this is the case it would be great so that vector won*t be impacted by this, because this behaviour only creates buggy logfiles but itself does not lead to a full filesystem.

**Comment by ulrichSchreiner** [comment]
hi, sorry for the late answer. here is the `managed.toml` content of the configmap. the code is generated, but i have masked some:
~~~
data_dir = "/vector-data-dir"

[api]
    enabled = false
    address = "0.0.0.0:8686"
    playground = true

[log_schema]
    host_key = "host"
    message_key = "message"
    source_type_key = "source_type"
    timestamp_key = "timestamp"

# Ingest logs from Kubernetes.
[sources.kubernetes_logs]
    type = "kubernetes_logs"

# Emit internal Vector metrics.
[sources.internal_metrics]
    type = "internal_metrics"

# Expose metrics for scraping in the Prometheus format.
[sinks.prometheus_sink]
    address = "0.0.0.0:9090"
    inputs = ["internal_metrics"]
    type = "prometheus"


[transforms.doorman-sample_json_log]
    inputs = ["ns_transformer.doorman-sample"]
    source = "s,err  = parse_json(.message)\nif err == null {\n  .log = s\n}\n.tenant = \"xxx\"\n.log_type = \"doorman\"\n.product_stage = \"test\"\n.app = .kubernetes.pod_labels.app\n"
    type = "remap"

[transforms.ingress-nginx_json_log]
    inputs = ["ns_transformer.ingress-nginx"]
    source = "s,err  = parse_json(.message)\nif err == null {\n  .log = s\n}\n.tenant = \"xxx\"\n.log_type = \"nginx\"\n"
    type = "remap"

[transforms.ns_transformer]
    inputs = ["kubernetes_logs"]
    type = "route"
    [transforms.ns_transformer.route]
        doorman-sample = ".kubernetes.pod_namespace == \"doorman-sample\""
        ingress-nginx = ".kubernetes.pod_namespace == \"ingress-nginx\""

[sinks.doorman-sample_loki_1]
    endpoint = "https://loki.xxxxx"
    inputs = ["doorman-sample_json_log"]
    out_of_order_action = "rewrite_timestamp"
    tenant_id = "xxx"
    type = "loki"
    [sinks.doorman-sample_loki_1.auth]
        password = "xxxxx"
        strategy = "basic"
        user = "yyyyy"
    [sinks.doorman-sample_loki_1.encoding]
        codec = "json"
    [sinks.doorman-sample_loki_1.healthcheck]
        enabled = true
    [sinks.doorman-sample_loki_1.labels]
        app = "{{ app }}"
        container = "{{ kubernetes.container_id }}"
        forwarder = "vector"
        log_type = "{{ log_type }}"
        pod = "{{ kubernetes.pod_name }}"
        product_stage = "{{ product_stage }}"
        stage = "test"
        tenant = "{{ tenant }}"

[sinks.ingress-nginx_loki_1]
    endpoint = "https://loki.xxxxx"
    inputs = ["ingress-nginx_json_log"]
    out_of_order_action = "rewrite_timestamp"
    tenant_id = "xxx"
    type = "loki"
    [sinks.ingress-nginx_loki_1.auth]
        password = "xxxxx"
        strategy = "basic"
        user = "yyyyy"
    [sinks.ingress-nginx_loki_1.encoding]
        codec = "json"
    [sinks.ingress-nginx_loki_1.healthcheck]
        enabled = true
    [sinks.ingress-nginx_loki_1.labels]
        app = "{{ app }}"
        container = "{{ kubernetes.container_id }}"
        forwarder = "vector"
        log_type = "{{ log_type }}"
        pod = "{{ kubernetes.pod_name }}"
        stage = "test"
        tenant = "{{ tenant }}"
~~~

**Comment by cmsxbc** [comment]
I have noticed you're using Loki. I suffered this issue before, and finally found that it's the output rate limit of loki sink. The default configuration of  Loki Sink is 100KiB or 100000 events in a batch, and at most 5 batch in 1 second. And the buffer default max_events is 500, it will block when it'full.

Increase `batch.max_bytes` in your configuration may inprove rate, and say good byte to this issue~

**Comment by jszwedko** [comment]
Hi @ulrichSchreiner ! A couple of thoughts:

Vector will hold the files open until it finishes reading all of the lines from them. My guess here is that Vector hasn't finished processing them. Once it has reached EOF, it will release the handle. This could be due to backpressure from the sink as @cmsxbc suggested where the sink can't send data fast enough to keep in with the incoming volume. If you are using the `loki` sink, it does have very low throughput currently as it is limited to one request at a time to avoid out-of-order issues in Loki. We have plans to improve that this quarter though.

> i'm not sure why these files are opened by the agent, because i thought only files which ends in *.log are monitored? are this files open because they were named 1.log and the rotation renamed them? in this case i expected the files to be closed.

There was a change in 0.15.0 to also monitor rotated files to reduce the chance that logs are lost in situations where the log volume spikes and Vector hasn't finished processing the original file before it is rotated. You can configure [`exclude_paths_glob_pattern`](https://vector.dev/docs/reference/configuration/sources/kubernetes_logs/#exclude_paths_glob_patterns) to exclude tailing them. Setting it to the default plus excluding rotated files should work: `["**/*.gz" "**/*.tmp", "**/*.log.*"]`

A question I have for you is whether you might prefer a different behavior where Vector releases the file handles when the files are deleted with the trade-off that any lines that Vector hasn't read yet will be lost. Would that be preferable in your situation?

**Comment by ulrichSchreiner** [comment]
hi,

as you asked my prefered behaviour: yes i'd like to release the handles when the files are deleted. in our case the files are rotated and only the third generation of old files are deleted. it is important for them to be deleted because the logs are on a system-volume of the k8s-nodes and if they are not deleted, the volumes runs up to 100% usage and the whole node is corrupt. so yes, in my case i'd like them to be released when k8s decides to delete the files. if i loose logs ... well that's sad but it is not an option to crash the whole node.

i'd also wanted to test the `batch.max_bytes` setting, but when i try to set this value in the helm values file, i came across the issue #7453 which is resolved by PR #8168. but when i try to use the `customConfig` with my current helm-values (which are rendered to a TOML file) i get a helm-error (i think i now have to escape many values like VRL code). so i sadly stumble from one problem to another ....

**Comment by jszwedko** [comment]
Hi @ulrichSchreiner . Thanks for clarifying! I opened https://github.com/timberio/vector/issues/8591 to add that enhancement.

I'll close this since I think it is represented by https://github.com/timberio/vector/issues/8591 but we can help troubleshoot your helm issue too if you like. I would recommend either opening a Github Discussion or dropping into [our discord server](https://chat.vector.dev/).
